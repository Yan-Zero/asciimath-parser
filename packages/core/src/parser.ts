import { TokenTypes } from './symbols'
import type { TokenizedValue } from './trie'

enum NodeTypes {
  Root = 'Root',
  Const = 'Const',
  ParamOne = 'ParamOne', // expression with one param
  ParamTwo = 'ParamTwo', // expression with two params
  Matrix = 'Matrix',
  Flat = 'Flat',
}

enum AlignDirection {
  Left = 'Left',
  Center = 'Center',
}

interface Node {
  type: NodeTypes
}

interface ConstNode extends Node {
  type: NodeTypes.Const
  value: string
  tex: string
}

interface RootNode extends Node {
  type: NodeTypes.Root
  body: ChildNode[]
}

interface FlatNode extends Node {
  type: NodeTypes.Flat
  body: ChildNode[]
}

interface ParamOneNode extends Node {
  type: NodeTypes.ParamOne
  params: ChildNode
  tex: string
}

interface ParamTwoNode extends Node {
  type: NodeTypes.ParamTwo
  params: [ChildNode, ChildNode]
  tex: string
}

interface MatrixNode extends Node {
  type: NodeTypes.Matrix
  params: ChildNode[][]
  lparen: string
  rparen: string
  alignment: AlignDirection
  dividerIndices: number[]
}

type ChildNode = ConstNode | MatrixNode | ParamOneNode | ParamTwoNode | FlatNode

function createRootNode(): RootNode {
  return {
    type: NodeTypes.Root,
    body: [],
  }
}
function createConstNode(): ConstNode
function createConstNode(tex: string): ConstNode
function createConstNode(token: TokenizedValue): ConstNode
function createConstNode(arg?: TokenizedValue | string) {
  if (typeof arg === 'undefined') {
    return {
      type: NodeTypes.Const,
      value: '',
      tex: '',
    }
  }
  if (typeof arg === 'string') {
    return {
      type: NodeTypes.Const,
      value: arg,
      tex: arg,
    }
  }
  return {
    type: NodeTypes.Const,
    value: arg.value,
    tex: arg.type === TokenTypes.Text ? `\\text{${arg.tex}}` : arg.tex,
  }
}

function createFlatNode(): FlatNode {
  return {
    type: NodeTypes.Flat,
    body: [],
  }
}

function createMatrixNode(): MatrixNode {
  return {
    type: NodeTypes.Matrix,
    params: [],
    lparen: '.',
    rparen: '.',
    alignment: AlignDirection.Center,
    dividerIndices: [],
  }
}

function createParamOneNode(): ParamOneNode {
  return {
    type: NodeTypes.ParamOne,
    tex: '',
    params: createFlatNode(),
  }
}

function createParamTwoNode(): ParamTwoNode {
  return {
    type: NodeTypes.ParamTwo,
    tex: '',
    params: [createFlatNode(), createFlatNode()],
  }
}

function createParenOfFlatNodeFrom(token: TokenizedValue, left: boolean): ConstNode {
  return { type: NodeTypes.Const, value: token.value, tex: `\\${left ? 'left' : 'right'}${token.tex}` }
}

function readParenedExpression2(tokens: TokenizedValue[], current: number): {
  node: MatrixNode | FlatNode
  current: number
} {
  const { closingIndex, semiIndex } = findPairedClosingParen(current, tokens)
  if (closingIndex === -1) {
    return parenStartedNoClosingNode(tokens, current)
  }
  else {
    if (semiIndex === -1) {
      // process the expression as an array
      return parenedArrayNode(tokens, current, closingIndex)
    }
    else {
      // process the expression as a matrix
      return generateMatrixNode(tokens, current, closingIndex)
    }
  }
}

function generateMatrixNode(tokens: TokenizedValue[], current: number, end: number) {
  let token = tokens[current]
  const node = createMatrixNode()
  const dividerIndices = new Set<number>()
  node.lparen = `\\left${token.tex}`
  token = tokens[++current]
  let tempArr: ChildNode[] = []
  let tempNode: ChildNode | null = null
  // inside a matrix
  while (current < end) {
    if (token.type === TokenTypes.Split) {
      if (token.value === ',') {
        if (tempNode) {
          tempArr.push(tempNode)
          tempNode = null
        }
        else {
          tempArr.push(createConstNode())
        }
      }
      else if (token.value === ';') {
        if (tempNode) {
          tempArr.push(tempNode)
          tempNode = null
        }
        node.params.push(tempArr)
        tempArr = []
      }
      token = tokens[++current]
      continue
    }
    else if (token.type === TokenTypes.Paren /* && token.value === '|' */) {
      if (tempNode) {
        tempArr.push(tempNode)
        tempNode = null
      }
      dividerIndices.add(tempArr.length)
      token = tokens[++current]
      continue
    }
    tempNode = createFlatNode()
    token = tokens[current]
    while (current < end
      && token.type !== TokenTypes.Split
      && token.type !== TokenTypes.Paren) {
      const walkRes = walk(tokens, current)
      current = walkRes.current
      tempNode.body.push(walkRes.node)
      token = tokens[current]
    }
  }
  // for those cases that the matrix only contains one line
  // or the last line of the matrix does not contains a semicolon
  if (tempNode) {
    tempArr.push(tempNode)
    tempNode = null
  }
  if (tempArr.length > 0) {
    node.params.push(tempArr)
    tempArr = []
  }
  // set dividerIndices
  node.dividerIndices = Array.from(dividerIndices).sort((a, b) => a - b)
  // process the right paren
  token = tokens[current]
  if (current < tokens.length) {
    current++
    node.rparen = `\\right${token.tex}`
    if (token.value === ':}')
      node.alignment = AlignDirection.Left
  }
  // unreachable
  else {
    // no right paren
    node.rparen = '\\right.'
  }
  return { node, current }
}

class ParenError extends Error {}

function parenedArrayNode(tokens: TokenizedValue[], current: number, closingIndex: number) {
  let token = tokens[current]
  const node = createFlatNode()
  node.body.push(createParenOfFlatNodeFrom(token, true))
  current = readTokensToFlatNode(current + 1, closingIndex, tokens, node)
  if (current >= tokens.length)
    throw new ParenError(`Read index out of range, index: ${current}`)

  token = tokens[current]
  current++
  node.body.push(createParenOfFlatNodeFrom(token, false))
  if ((node.body[0] as ConstNode).value === '{:' && (node.body[node.body.length - 1] as ConstNode).value === ':}') {
    (node.body[0] as ConstNode).tex = '{';
    (node.body[node.body.length - 1] as ConstNode).tex = '}'
  }
  return { node, current }
}

function parenStartedNoClosingNode(tokens: TokenizedValue[], current: number) {
  const token = tokens[current]
  // just process the expression as flat one
  const node = createFlatNode()
  node.body.push({ type: NodeTypes.Const, value: token.value, tex: `\\left${token.tex}` } as ConstNode)
  current = readTokensToFlatNode(current + 1, tokens.length, tokens, node)
  // maybe it's better to add a hidden closing bracket
  node.body.push({ type: NodeTypes.Const, value: token.value, tex: '\\right.' } as ConstNode)
  return { node, current }
}

function findPairedClosingParen(current: number, tokens: TokenizedValue[]) {
  let semiIndex = -1
  let closingIndex = -1
  const stack: string[] = []
  // find the paired closing paren and whether there is any `;` within the parens
  for (let i = current + 1; i < tokens.length; i++) {
    if (tokens[i].type === TokenTypes.LParen) {
      stack.push('')
      continue
    }
    if (stack.length === 0) {
      if (tokens[i].value === ';') {
        if (semiIndex === -1)
          semiIndex = i
      }
      else if (tokens[i].type === TokenTypes.RParen) {
        if (closingIndex === -1)
          closingIndex = i
      }
      if (semiIndex !== -1 && closingIndex !== -1)
        break
    }
    else {
      if (tokens[i].type === TokenTypes.RParen)
        stack.pop()
    }
  }
  return { closingIndex, semiIndex }
}

function findPairedBar(arr: TokenizedValue[], start: number, end: number): {
  semiIndex: number
  barIndex: number
} {
  let semiIndex = -1
  let barIndex = -1
  const stack: string[] = []
  for (let i = start; i < end; i++) {
    if (arr[i].type === TokenTypes.LParen) {
      stack.push('')
      continue
    }
    if (stack.length > 0 && arr[i].type === TokenTypes.RParen) {
      stack.pop()
      continue
    }
    if (stack.length > 0)
      continue
    if (arr[i].type === TokenTypes.RParen)
      break
    if (arr[i].value === ';') {
      if (semiIndex === -1)
        semiIndex = i
    }
    else if (arr[i].value === '|') {
      if (barIndex === -1)
        barIndex = i
    }
    if (semiIndex !== -1 && barIndex !== -1)
      break
  }
  return { semiIndex, barIndex }
}

function readBarStartedExpressions(tokens: TokenizedValue[], current: number): {
  node: ChildNode
  current: number
} {
  let token = tokens[current]
  // in fact only `|` matches `TokenTypes.Paren`
  const { semiIndex, barIndex } = findPairedBar(tokens, current + 1, tokens.length)

  /**
     * case 1: not matrix
     * there isn't any `;` split pattern before a Paren or a RParen
     * for example `{ (x, y) | x^2 + y^2 <= 1 }`
     */
  if (barIndex === -1)
    return createSingleBarNode(current)

  // used as `abs`
  if (semiIndex === -1 || semiIndex > barIndex) {
    const node = createFlatNode()
    current++
    node.body.push(createConstNode('\\left|'))
    current = readTokensToFlatNode(current, barIndex, tokens, node)
    node.body.push(createConstNode('\\right|'))
    current = barIndex + 1
    return { current, node }
  }

  /**
     * case 2: matrix
     * there is one or more than one `;` before a Paren ~~or a RParen~~
     *
     * common cases:
     * `| 1, 2; |` or `| 1, 2; 3, 4 |`
     *
     * special case:
     * `| 1, 2, | 3, 4; 5, 6 |, 4; ... |`
     * maybe I should use a `stack` to store the parens
     * however, the case is too complex, so I won't implement it ¯\_(ツ)_/¯
     */
  const node = createMatrixNode()
  node.lparen = '\\left|'
  node.rparen = '\\right|'

  token = tokens[++current]
  let tempArr: ChildNode[] = []
  let tempNode: ChildNode | null = null
  while (current < barIndex) {
    if (token.type === TokenTypes.Split) {
      switch (token.value) {
        case ',': {
          if (tempNode) {
            tempArr.push(tempNode)
            tempNode = null
          }
          break
        }
        case ';': {
          if (tempNode) {
            // the ownership of `tempNode` is passed to `tempArr`
            tempArr.push(tempNode)
            tempNode = null
          }
          // the ownership of `tempArr` is passed to `node.params`
          node.params.push(tempArr)
          tempArr = []
          break
        }
      }
      token = tokens[++current]
      continue
    }
    tempNode = createFlatNode()
    token = tokens[current]
    while (current < barIndex
        && token.type !== TokenTypes.Split) {
      const walkRes = walk(tokens, current)
      current = walkRes.current
      tempNode.body.push(walkRes.node)
      token = tokens[current]
    }
  }
  if (tempNode) {
    tempArr.push(tempNode)
    tempNode = null
  }
  if (tempArr.length > 0) {
    node.params.push(tempArr)
    tempArr = []
  }
  current = barIndex + 1
  return { node, current }
}

function readTokensToFlatNode(current: number, end: number, tokens: TokenizedValue[], node: FlatNode) {
  while (current < end) {
    const walkRes = walk(tokens, current)
    current = walkRes.current
    node.body.push(walkRes.node)
  }
  return current
}

function createSingleBarNode(current: number): { node: ChildNode; current: number } {
  return {
    current: current + 1,
    node: {
      type: NodeTypes.Const,
      value: '|',
      tex: '\\mid',
    },
  }
}

function removeParenOfFlatExpr(node: FlatNode): FlatNode {
  const first = node.body[0]
  const last = node.body[node.body.length - 1]
  if (first.type === NodeTypes.Const
    && last.type === NodeTypes.Const
    && (first as ConstNode).value === '('
    && (last as ConstNode).value === ')') {
    node.body.pop()
    node.body.shift()
  }
  return node
}

function lookForwardOperatorOptionalTwoParams(tokens: TokenizedValue[], current: number, token: TokenizedValue) {
  let nextToken = tokens[current]
  let p1: ChildNode = createConstNode()
  let nextShouldBe = ''
  let p1Status = ''
  let p2Status = ''
  if (nextToken.value === '^' || nextToken.value === '_') {
    nextShouldBe = nextToken.value === '^' ? '_' : '^'
    p1Status = nextToken.value
    current++
    // do not look forward
    // do not receive other sup or sub in recursion
    const walkRes = walk(tokens, current, false)
    if (walkRes.node.type === NodeTypes.Flat)
      walkRes.node = removeParenOfFlatExpr(walkRes.node)
    p1 = walkRes.node
    current = walkRes.current
  }
  let p2: ChildNode = createConstNode()
  if (current < tokens.length) {
    nextToken = tokens[current]
    if (nextToken.value === nextShouldBe) {
      p2Status = nextToken.value
      current++
      // do not look forward
      // do not receive other sup or sub in recursion
      const walkRes = walk(tokens, current, false)
      if (walkRes.node.type === NodeTypes.Flat)
        walkRes.node = removeParenOfFlatExpr(walkRes.node)
      p2 = walkRes.node
      current = walkRes.current
    }
  }
  const node = createParamTwoNode()
  node.tex = token.tex
  node.params[0] = (() => {
    if (p1Status === '^')
      return p1
    if (p2Status === '^')
      return p2
    return createConstNode()
  })()
  node.params[1] = (() => {
    if (p1Status === '_')
      return p1
    if (p2Status === '_')
      return p2
    return createConstNode()
  })()
  return { node, current }
}

function preProcessOperatorSup(node: ChildNode, operator: TokenizedValue, tokens: TokenizedValue[], current: number): WalkResult {
  let newNode: FlatNode
  if (node.type === NodeTypes.Flat) {
    newNode = node
  }
  else {
    newNode = createFlatNode()
    newNode.body.push(node)
  }
  const supNode = createParamOneNode()
  supNode.tex = operator.tex
  // do not look forward
  // avoid receiving fractions or other nodes as sup node
  const walkRes = walk(tokens, current, false)
  current = walkRes.current
  if (walkRes.node.type === NodeTypes.Flat)
    walkRes.node = removeParenOfFlatExpr(walkRes.node)
  supNode.params = walkRes.node
  newNode.body.push(supNode)
  node = newNode
  return { node, current }
}

function preProcessOperatorAO(node: ChildNode, nextToken: TokenizedValue) {
  const newNode = createFlatNode()
  if (node.type === NodeTypes.Flat)
    newNode.body.push(...(node.body))

  else
    newNode.body.push(node)
  newNode.body.push(createConstNode(nextToken))
  node = newNode
  return node
}

function preProcessOperatorAOB(node: ChildNode, operator: TokenizedValue, tokens: TokenizedValue[], current: number): WalkResult {
  const newNode = createParamTwoNode()
  if (node.type === NodeTypes.Flat)
    node = removeParenOfFlatExpr(node)
  newNode.tex = operator.tex
  newNode.params[0] = node
  const walkRes = walk(tokens, current)
  current = walkRes.current
  if (walkRes.node.type === NodeTypes.Flat)
    walkRes.node = removeParenOfFlatExpr(walkRes.node)
  newNode.params[1] = walkRes.node
  node = newNode
  return { node, current }
}

function getParamOneNode(tokens: TokenizedValue[], current: number, lookForward: boolean): { node: ParamOneNode; current: number } {
  const node = createParamOneNode()
  const token = tokens[current]
  node.tex = token.tex
  current++
  const walkRes = walk(tokens, current, lookForward)
  current = walkRes.current
  if (walkRes.node.type === NodeTypes.Flat)
    walkRes.node = removeParenOfFlatExpr(walkRes.node)

  node.params = walkRes.node
  return { node, current }
}

interface WalkResult {
  node: ChildNode
  current: number
}

function walk(tokens: TokenizedValue[], current: number, watchNext = true): WalkResult {
  if (current >= tokens.length)
    return { node: createConstNode(), current }
  const token = tokens[current]
  let node: ChildNode
  switch (token.type) {
    case TokenTypes.Const:
    case TokenTypes.Text:
    case TokenTypes.NumberLiteral:
    case TokenTypes.StringLiteral: {
      current++
      node = createConstNode(token)
      break
    }
    case TokenTypes.LParen: {
      ({ node, current } = readParenedExpression2(tokens, current))
      break
    }
    case TokenTypes.Paren: {
      ({ node, current } = readBarStartedExpressions(tokens, current))
      break
    }
    case TokenTypes.OperatorSup:
    case TokenTypes.OperatorA: {
      // high priority, do not look forward
      ({ node, current } = getParamOneNode(tokens, current, false))
      break
    }
    case TokenTypes.OperatorMinus: {
      // low priority, do look forward
      ({ node, current } = getParamOneNode(tokens, current, true))
      break
    }
    case TokenTypes.OperatorOAB: {
      node = createParamTwoNode()
      node.tex = token.tex
      current++
      const param0 = walk(tokens, current)
      current = param0.current
      if (param0.node.type === NodeTypes.Flat)
        param0.node = removeParenOfFlatExpr(param0.node)
      node.params[0] = param0.node
      const param1 = walk(tokens, current)
      current = param1.current
      if (param1.node.type === NodeTypes.Flat)
        param1.node = removeParenOfFlatExpr(param1.node)
      node.params[1] = param1.node
      break
    }
    case TokenTypes.OperatorO2: {
      current++
      if (current >= tokens.length) {
        node = createConstNode(`${token.tex.replace(/[\{\[] \$\d+ [\}\]]/g, '')}{}`)
        break
      }

      // detect next token
      ({ node, current } = lookForwardOperatorOptionalTwoParams(tokens, current, token))
      break
    }
    case TokenTypes.Split:
    case TokenTypes.Align: {
      current++
      node = createConstNode(token)
      break
    }
    case TokenTypes.RParen: {
      current++
      node = createConstNode(token)
      break
    }
    default: {
      throw new Error(`Unmatched token in walk ${token.value}`)
    }
  }
  // watch next token
  if (current < tokens.length && watchNext) {
    let matched = true
    while (matched && current < tokens.length) {
      const nextToken = tokens[current]
      switch (nextToken.type) {
        case TokenTypes.OperatorAOB: {
          ({ node, current } = preProcessOperatorAOB(node, nextToken, tokens, current + 1))
          break
        }
        case TokenTypes.OperatorAO: {
          node = preProcessOperatorAO(node, nextToken)
          current++
          break
        }
        case TokenTypes.OperatorSup: {
          ({ node, current } = preProcessOperatorSup(node, nextToken, tokens, current + 1))
          break
        }
        default: {
          matched = false
        }
      }
    }
  }
  return { node, current }
}

function parser(tokens: TokenizedValue[]) {
  const root = createRootNode()
  let current = 0

  while (current < tokens.length) {
    const walkRes = walk(tokens, current)
    current = walkRes.current
    root.body.push(walkRes.node)
  }

  return root
}

export {
  NodeTypes,
  Node,
  ChildNode,
  RootNode,
  AlignDirection,
  MatrixNode,
  parser,
}
