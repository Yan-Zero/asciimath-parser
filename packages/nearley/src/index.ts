import Nearley from 'nearley'
import initGrammar from './grammar.js'
import initLexer from './lexer.js'
import type { Symbols } from './symbols.js'
import initSymbols from './symbols.js'
import initGenerator from './to-tex.js'

export type Ast = any
type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * @default false
   * throws error on parse
   */
  throws?: boolean
  /**
   * Add custom symbols.
   *
   * For example:
   * {
   *   keyword: {
   *     dx: { tex: '{\\text{d}x}' },
   *     dy: { tex: '{\\text{d}y}' },
   *     dz: { tex: '{\\text{d}z}' },
   *     dt: { tex: '{\\text{d}t}' },
   *     ee: { tex: '\\text{e}' },
   *     ii: { tex: '\\text{i}' },
   *   }
   * }
   */
  symbols?: Symbols
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  return {
    display: config?.display ?? true,
    throws: config?.throws ?? false,
    symbols: {
      keyword: {
        dx: { tex: '{\\text{d}x}' },
        dy: { tex: '{\\text{d}y}' },
        dz: { tex: '{\\text{d}z}' },
        dt: { tex: '{\\text{d}t}' },
        ee: { tex: '\\text{e}' },
        ii: { tex: '\\text{i}' },
      },
      ...config?.symbols,
    },
    replaceBeforeTokenizing: [
      // html entity
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
      ...(config?.replaceBeforeTokenizing || []),
    ],
  }
}

class AsciiMath {
  public display: boolean
  public throws: boolean
  public replaceBeforeTokenizing: ReplaceLaw[]
  public lexer: Nearley.Lexer
  public parser: Nearley.Parser
  private genTex: ((ast: Ast) => string)
  private initState: { [key: string]: any; lexerState: Nearley.LexerState }
  constructor(config?: AsciiMathConfig) {
    const { display, throws, symbols, replaceBeforeTokenizing } = resolveConfig(config)
    this.display = display
    this.throws = throws
    this.replaceBeforeTokenizing = replaceBeforeTokenizing
    const allSymbols = initSymbols(symbols)
    const lexer = this.lexer = initLexer(allSymbols)
    const grammar = initGrammar(lexer)
    const compiledGrammar = Nearley.Grammar.fromCompiled(grammar)
    this.parser = new Nearley.Parser(compiledGrammar)
    this.initState = this.parser.save()
    this.genTex = initGenerator(allSymbols)
  }

  toTex(code: string): string {
    this.parser.restore(this.initState)
    try {
      code = this.replaceBeforeTokenizing.reduce((prev, curLaw) => {
        // nodejs 14 没有 replaceAll 方法, 不要用 replaceAll
        return prev.replace(new RegExp(curLaw[0], 'g'), curLaw[1] as any)
      }, code)
      this.parser.feed(code)
      if (this.parser.results.length === 0) {
        throw new Error('unexpected end of input')
      }
      else if (this.parser.results.length > 1) {
        console.error(this.parser.results)
        throw new Error('ambiguous parse')
      }
      let res = this.genTex(this.parser.results)
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      if (this.throws)
        throw e
      console.error(e)
      const message = String(e)
      let index = message.indexOf(' Instead, I was expecting to see one of the following:')
      if (index > -1) {
        return this.genTex({
          type: 'opOA',
          value: 'verb',
          $1: { value: message.slice(0, index) },
        })
      }

      index = message.indexOf(':\n')
      const end = index > -1 ? index : undefined
      return `\\text{${message.slice(0, end)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
