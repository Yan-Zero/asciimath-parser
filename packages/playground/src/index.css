@import url('https://fonts.googleapis.com/css2?family=JetBrains+Mono&display=swap');

:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}

body {
  margin: 0;
  min-width: 320px;
  min-height: 100vh;
}

h1 {
  font-size: 3.2em;
  line-height: 1.1;
  margin-block-end: 0.25rem;
}

h2 {
  margin-top: 3rem;
}

h3 {
  margin-top: 2rem;
}

button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
  display: flex;
  align-items: center;
  justify-content: center;
}
button:hover {
  border-color: #646cff;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  a:hover {
    color: #747bff;
  }
  button {
    background-color: #f9f9f9;
  }
}

td {
  padding: 0.5rem;
  white-space: nowrap;
}

table {
  width: 100%;
  border-top: 2px solid;
  border-bottom: 2px solid;
}

.tc {
  text-align: center;
}

.tex {
  background-color: #7f7f7f33;
}

main {
  max-width: 800px;
  margin: 0 auto;
  padding: 1rem;
}

.space-x-1 > :not([hidden]) ~ :not([hidden]) {
  --tw-space-x-reverse: 0;
  margin-right: calc(0.25rem * var(--tw-space-x-reverse));
  margin-left: calc(0.25rem * calc(1 - var(--tw-space-x-reverse)));
}

.flex {
  display: flex;
}
.w-full {
  width: 100%;
}
.mb-4 {
  margin-bottom: 1rem;
}

code, pre {
  font-family: 'JetBrains Mono', 'Courier New', Courier, monospace;
  font-size: 0.8rem;
}

code {
  background-color: #7f7f7f11;
  padding: 0.2rem;
  border-radius: 0.2rem;
}

pre {
  background-color: #7f7f7f11;
  padding: 0.75em;
  border-radius: 0.2rem;
  overflow: auto;
}