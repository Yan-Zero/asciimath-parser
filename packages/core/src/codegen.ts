import type { ChildNode, MatrixNode, RootNode } from './parser'
import { AlignDirection, NodeTypes } from './parser'

function getArrayBoundary(node: MatrixNode) {
  const div = node.dividerIndices
  let beginArray = '\\begin{array}'
  let ch = 'c'
  if (node.alignment === AlignDirection.Left)
    ch = 'l'

  if (div.length) {
    const divMax = div[div.length - 1]
    for (let i = div.length - 1; i >= 1; i--)
      div[i] -= div[i - 1]

    beginArray += '{'
    for (let i = 0; i < div.length; i++)
      beginArray += `${''.padEnd(div[i], ch)}|`

    // MathJax would complain if the array env arg
    // is not consistent with the elements of the matrix.
    const maxCol = Math.max(...node.params.map(i => i.length))
    beginArray += `${''.padEnd(maxCol - divMax, ch)}}`
  }
  else {
    const maxCol = Math.max(...node.params.map(c => c.length))
    beginArray += '{' + ''.padEnd(maxCol, ch) + '}'
  }
  return [beginArray, '\\end{array}']
}

function codegen(node: ChildNode | RootNode): string {
  switch (node.type) {
    case NodeTypes.Const: {
      return node.tex
    }
    case NodeTypes.Root: {
      let res = node.body.map(codegen).join(' ')
      if (node.body.find(n => n.type === NodeTypes.Const && n.value === '&'))
        res = `\\begin{aligned}${res}\\end{aligned}`
      return res
    }
    case NodeTypes.Flat: {
      return node.body.map(codegen).join(' ')
    }
    case NodeTypes.Matrix: {
      // const [beginMatrix, endMatrix] = getMatrixBoundary(node)
      const [arrayBegin, arrayEnd] = getArrayBoundary(node)
      return [
        node.lparen,
        // beginMatrix,
        arrayBegin,
        node.params.map(i => i.map(codegen).join('&')).join('\\\\'),
        '\\\\',
        arrayEnd,
        // endMatrix,
        node.rparen,
      ].join(' ')
    }
    case NodeTypes.ParamOne: {
      return node.tex.replace('$1', codegen(node.params))
    }
    case NodeTypes.ParamTwo: {
      return node.tex.replace('$1', codegen(node.params[0])).replace('$2', codegen(node.params[1]))
    }
  }
}

export {
  codegen,
}
