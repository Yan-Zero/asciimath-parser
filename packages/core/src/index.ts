import { codegen } from './codegen'
import { parser } from './parser'
import type { Trie } from './trie'
import { createTrie } from './trie'

type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * Translate custom keywords into LaTeX expressions
   *
   * For example:
   * [
   *   ['dx', '\text{d}x'],
   *   ['dy', '\text{d}y']
   * ]
   */
  extConst?: Array<[string, string]>
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  const defaultConfig: RestrictedAmConfig = {
    display: true,
    extConst: [
      ['dx', '{\\text{d}x}'],
      ['dy', '{\\text{d}y}'],
      ['dz', '{\\text{d}z}'],
      ['dt', '{\\text{d}t}'],
      ['#', '\\displaystyle'],
    ],
    replaceBeforeTokenizing: [
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
    ],
  }
  if (typeof config?.display !== 'undefined')
    defaultConfig.display = config?.display
  if (config?.extConst?.length)
    defaultConfig.extConst.push(...config.extConst)
  if (config?.replaceBeforeTokenizing?.length)
    defaultConfig.replaceBeforeTokenizing.push(...config.replaceBeforeTokenizing)

  return defaultConfig
}

class AsciiMath {
  private trie: Trie
  private display: boolean
  private replaceLaws: ReplaceLaw[]
  constructor(config?: AsciiMathConfig) {
    const { display, extConst, replaceBeforeTokenizing: replaceBeforeParsing } = resolveConfig(config)
    this.trie = createTrie({ extConst })
    this.display = display
    this.replaceLaws = replaceBeforeParsing
  }

  toTex(code: string): string {
    try {
      code = this.replaceLaws.reduce((prev, curLaw) => {
      // in fact the judgement is no use...
        if (typeof curLaw[1] === 'function')
          return prev.replace(curLaw[0], curLaw[1])
        else
          return prev.replace(curLaw[0], curLaw[1])
      }, code)
      let res = codegen(parser(this.trie.tryParsingAll(code)))
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      return `\\text{${String(e)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
