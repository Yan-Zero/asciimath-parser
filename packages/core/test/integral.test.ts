import { describe, expect, it } from 'vitest'
import { AsciiMath } from '../src'
import { createTrie } from '../src/trie'
import { parser } from '../src/parser'
import { codegen } from '../src/codegen'

describe('asciimath', () => {
  it('should parse integrals', () => {
    const trie = createTrie()
    const code = 'int "e"^-x dx '
    const tokens = trie.tryParsingAll(code)
    expect(tokens).toMatchSnapshot()
    const ast = parser(tokens)
    expect(ast).toMatchSnapshot()
    const res = codegen(ast)
    expect(res).toMatchSnapshot()
  })
})
