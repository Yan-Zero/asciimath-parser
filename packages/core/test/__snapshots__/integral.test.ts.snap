// Vitest Snapshot v1

exports[`asciimath > should parse integrals 1`] = `
[
  {
    "current": 3,
    "isKeyWord": true,
    "tex": "\\\\int",
    "type": "Const",
    "value": "int",
  },
  {
    "current": 7,
    "isKeyWord": false,
    "tex": "e",
    "type": "Text",
    "value": "e",
  },
  {
    "current": 8,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 9,
    "isKeyWord": true,
    "tex": "{-$1 }",
    "type": "OperatorMinus",
    "value": "-",
  },
  {
    "current": 10,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 12,
    "isKeyWord": false,
    "tex": "d",
    "type": "StringLiteral",
    "value": "d",
  },
  {
    "current": 13,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
]
`;

exports[`asciimath > should parse integrals 2`] = `
{
  "body": [
    {
      "tex": "\\\\int",
      "type": "Const",
      "value": "int",
    },
    {
      "body": [
        {
          "tex": "\\\\text{e}",
          "type": "Const",
          "value": "e",
        },
        {
          "params": {
            "params": {
              "tex": "x",
              "type": "Const",
              "value": "x",
            },
            "tex": "{-$1 }",
            "type": "ParamOne",
          },
          "tex": "^{ $1 }",
          "type": "ParamOne",
        },
      ],
      "type": "Flat",
    },
    {
      "tex": "d",
      "type": "Const",
      "value": "d",
    },
    {
      "tex": "x",
      "type": "Const",
      "value": "x",
    },
  ],
  "type": "Root",
}
`;

exports[`asciimath > should parse integrals 3`] = `"\\\\int \\\\text{e} ^{ {-x } } d x"`;
